File: survival-client/src/main/java/net/survival/client/input/Keyboard.java
package net.survival.client.input;

public final class Keyboard
{
    static boolean[] prevKeyState = new boolean[Key.getCachedValues().length];
    static boolean[] keyState = new boolean[Key.getCachedValues().length];

    private Keyboard() {}

    public static boolean isKeyDown(Key key) {
        return keyState[key.ordinal()];
    }

    public static boolean isKeyUp(Key key) {
        return !keyState[key.ordinal()];
    }

    public static boolean isKeyPressed(Key key) {
        return keyState[key.ordinal()] && !prevKeyState[key.ordinal()];
    }

    public static boolean isKeyReleased(Key key) {
        return !keyState[key.ordinal()] && prevKeyState[key.ordinal()];
    }
}

File: survival-client/src/main/java/net/survival/client/input/Keyboard.kt
package net.survival.client.input

object Keyboard {
    var prevKeyState = BooleanArray(Key.cachedValues.size)
    var keyState = BooleanArray(Key.cachedValues.size)

    fun isKeyDown(key: Key): Boolean = keyState[key.ordinal]
    fun isKeyUp(key: Key): Boolean = !keyState[key.ordinal]

    fun isKeyPressed(key: Key): Boolean =
        keyState[key.ordinal] && !prevKeyState[key.ordinal]

    fun isKeyReleased(key: Key): Boolean =
        !keyState[key.ordinal] && prevKeyState[key.ordinal]
}